### PR TITLE
Add a way to create a application credentials json file

### PR DESCRIPTION
**Pull request by GregoireW**
Sometime it is easier to use authentication helpers to manage authentication as it means same code for different scenario (like running a process in your computer, on a CI or on production )

I added this option. In a way, it is not really linked to the core of this action, but as it is really really close, I feel it can be easier to have this option here.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,2692 +1,8 @@
 {
   "name": "oidc-auth-gcp",
   "version": "0.1.0",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Comment (sethvargo) [inline package-lock.json:4]: This lockfile is important. Please revert these changes.

File: src/main.ts [hunk 2]
```diff
@@ -24,6 +27,40 @@ function explodeStrings(input: string): Array<string> {
...
+  workloadIdentityProvider: string,
+  serviceAccount: string,
+  githubOIDCToken: string,
+) {
+  const rootDir = process.env.GITHUB_WORKSPACE;
+  if (!rootDir) {
+    throw new Error('No path for credentials. Set process.env.GITHUB_WORKSPACE');
+  }
+  // Write github OIDC token
+  const oidcTokenPath = path.join(rootDir, uuidv4());
```
Comment (sethvargo) [inline src/main.ts:40]: We should not use UUIDs for filepaths since it brings in an unnecessary external dependency and still has a chance of conflicting. We should use the primitives in the `fs` package for making a temporary file, since those will automatically retry on conflicts.

File: src/main.ts [hunk 3]
```diff
@@ -45,6 +82,11 @@ async function run(): Promise<void> {
     // Get the GitHub OIDC token.
     const githubOIDCToken = await core.getIDToken(audience);
 
+    if (tokenFormat === 'application_credentials') {
```
Comment (sethvargo) [inline src/main.ts:85]: I don't think "application_credentials" is a token format. This doesn't really make sense.

File: src/main.ts [hunk 4]
```diff
@@ -68,6 +110,9 @@ async function run(): Promise<void> {
         break;
       }
       case 'id_token': {
+        if (idTokenAudience === '') {
```
Comment (sethvargo) [inline src/main.ts:113]: I prefer to let this flow through to the API which will reject it instead of doing client-side validation. The upstream API could change and then the client is too restrictive, which isn't great.

**Review by sethvargo — CHANGES_REQUESTED** [review]
Hi @GregoireW - thank you for opening a Pull Request. I don't think this can be merged as-is, and I've suggested a few changes. I also replied to your issue about how to best authenticate to Docker using this action, and it does not require this credentials file.

**Comment by GregoireW** [comment]
I knew this PR would not be correct the first try as I wasn't 100% satisfied with it, but thank you to correct it by yourself by integrating it to the core action! 

Note:  I added the `if (idTokenAudience === '') { `  because the API (hence the action) return a `400 bad request` error without specifying the audience is mandatory.  It is a little bit frustrating because you may not be 100% sure of your workload identity setup and you tend to update this setup instead of the action configuration.   Perhaps adding a code in the catch(error)  like ' if token is idtoken and audience is not given then print "You may have forgotten to set audience" '  is a good compromise.
